### PR TITLE
feat: (`SyncHighestDecided`) wait for minimum peers before requesting highest decided

### PR DESCRIPTION
**Pull request by moshe-blox**

### CODE REVIEW

File: network/p2p/p2p_sync.go [hunk 1]
```diff
@@ -75,8 +73,8 @@ func (n *p2pNetwork) LastDecided(logger *zap.Logger, mid spectypes.MessageID) ([
 	if !n.isReady() {
 		return nil, p2pprotocol.ErrNetworkIsNotReady
 	}
-	pid, peerCount := n.fork.ProtocolID(p2pprotocol.LastDecidedProtocol)
-	peers, err := n.getSubsetOfPeers(logger, mid.GetPubKey(), peerCount, allPeersFilter)
+	pid, maxPeers := n.fork.ProtocolID(p2pprotocol.LastDecidedProtocol)
+	peers, err := waitSubsetOfPeers(logger, n.getSubsetOfPeers, mid.GetPubKey(), 3, maxPeers, time.Second*24, allPeersFilter)
```
Comment (y0sher) [inline network/p2p/p2p_sync.go:77]: IMO `3` and `time.Second*24` should be consts or at least vars, solely for the purpose of them being named.

File: network/p2p/p2p_sync.go [hunk 2]
```diff
@@ -300,3 +288,42 @@ func (n *p2pNetwork) peersWithProtocolsFilter(protocols ...libp2p_protocol.ID) f
...
+	}
+	if minPeers < 0 || maxPeers < 0 {
+		return nil, fmt.Errorf("minPeers and maxPeers should not be negative")
+	}
+	if timeout <= 0 {
+		return nil, fmt.Errorf("timeout should be positive")
+	}
+
+	// Wait for minPeers with a deadline.
+	deadline := time.Now().Add(timeout)
```
Comment (y0sher) [inline network/p2p/p2p_sync.go:311]: could've been done with context.Deadline as well. though I don't see a real issue with current approach. once we'll want to introduce shutdown we'll need to pass a context down and combine it with this.

File: network/p2p/p2p_sync.go [hunk 3]
```diff
@@ -300,3 +288,42 @@ func (n *p2pNetwork) peersWithProtocolsFilter(protocols ...libp2p_protocol.ID) f
...
+	for {
+		peers, err := getSubsetOfPeers(logger, vpk, maxPeers, filter)
+		if err != nil {
+			return nil, err
+		}
+		if len(peers) >= minPeers || minPeers == 0 {
```
Comment (y0sher) [inline network/p2p/p2p_sync.go:317]: I don't get why we need to `minPeers == 0` condition if it is never used.

File: network/p2p/p2p_sync.go (hunk 2 above)
Comment (moshe-blox) [reply network/p2p/p2p_sync.go:311]: this func doesn't accept context (like all other funcs in this struct)

what needs to be done is a huge refactor to every blocking func/method in SSV to accept context and pass it down correctly

though, for graceful shutdown the only important piece is the database (imo), so we can make that refactor a bit smaller

File: network/p2p/p2p_sync.go (hunk 3 above)
Comment (moshe-blox) [reply network/p2p/p2p_sync.go]: leftover from another condition — deleted

**Review by y0sher — APPROVED** [review]

**Review by lior-blox — APPROVED** [review]

**Pull request merged**